aten/src/ATen/core/intrusive_ptr.h

```diff
@@ -0,0 +1,532 @@
+   * destruct it yet because we need the weakcount accessible).
+   *
+   * Even if there are no weak references (i.e. your class is about to be
+   * destructed), this function is guaranteed to be called first.
+   * However, if you use your class for an object on the stack that is
+   * destructed by the scope (i.e. without intrusive_ptr), this function will
+   * not be called.
+   */
+  virtual void free() const {}
+
+  template <class TTarget, class NullType>
+  friend class intrusive_ptr;
+};
+
+namespace detail {
+template <class TTarget>
+struct intrusive_target_default_null_type final {
+  static constexpr TTarget* singleton() noexcept {
+    return nullptr;
+  }
+};
+} // namespace detail
+
+template <class TTarget, class NullType>
+class weak_intrusive_ptr;
+
+template <
+    class TTarget,
+    class NullType = detail::intrusive_target_default_null_type<TTarget>>
+class intrusive_ptr final {
+ private:
+  static_assert(
+      std::is_base_of<intrusive_ptr_target, TTarget>::value,
+      "intrusive_ptr can only be used for classes that inherit from intrusive_ptr_target.");
+  static_assert(
+      NullType::singleton() == NullType::singleton(),
+      "NullType must have a constexpr singleton() method");
+  static_assert(
+      std::is_same<TTarget*, decltype(NullType::singleton())>::value,
+      "NullType::singleton() must return a element_type* pointer");
+
+  TTarget* target_;
+
+  template <class TTarget2, class NullType2>
+  friend class intrusive_ptr;
+  friend class weak_intrusive_ptr<TTarget, NullType>;
+
+  void retain() noexcept {
+    if (target_ != NullType::singleton()) {
+      size_t new_refcount = ++target_->refcount_;
+      AT_CHECK(
+          new_refcount !=
+          1) // intrusive_ptr: Cannot increase refcount after it reached zero.
+    }
+  }
+
+  void release() noexcept {
+    if (target_ != NullType::singleton() && --target_->refcount_ == 0) {
+      // See comment above about weakcount. As long as refcount>0,
+      // weakcount is one larger than the actual number of weak references.
+      // So we need to decrement it here.
+      auto weak_count = --target_->weakcount_;
+      target_->free();
+      if (weak_count == 0) {
+        delete target_;
+      }
+    }
+    target_ = NullType::singleton();
+  }
+
+  // This constructor will not increase the ref counter for you.
+  // This is not public because we shouldn't make intrusive_ptr out of raw
+  // pointers except from inside the make_intrusive() and
+  // weak_intrusive_ptr::lock() implementations
+  explicit intrusive_ptr(TTarget* target) noexcept : target_(target) {}
+
+ public:
+  using element_type = TTarget;
+
+  intrusive_ptr() noexcept : intrusive_ptr(NullType::singleton()) {}
+
+  intrusive_ptr(intrusive_ptr&& rhs) noexcept : target_(rhs.target_) {
+    rhs.target_ = NullType::singleton();
+  }
+
+  template <class From, class FromNullType>
+  intrusive_ptr(intrusive_ptr<From, FromNullType>&& rhs) noexcept
```

Comment:
What do we need the convertible constructor for?

Reply:
converting `intrusive_ptr<ChildClass>` to `intrusive_ptr<BaseClass>`